### PR TITLE
63 include custom filter in experimental

### PR DESCRIPTION
**Pull request by frederik-sandfort1**
Please have a look @c-w-feldmann and @JochenSiegWork  :-)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (c-w-feldmann) [inline .vscode/settings.json:1]: please remove the IDE settings

File: tests/test_elements/test_experimental/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""Init."""
```
Comment (c-w-feldmann) [inline tests/test_elements/test_experimental/__init__.py:1]: """Initialize module for experimental classes and functions."""

**Review by c-w-feldmann — CHANGES_REQUESTED** [review]
Minor changes

File: .vscode/settings.json (hunk 1 above)
Comment (frederik-sandfort1) [reply .vscode/settings.json:1]: sorry for this one... I always expect this to be in .gitignore... will add ;-)

File: tests/test_elements/test_experimental/__init__.py (hunk 2 above)
Comment (frederik-sandfort1) [reply tests/test_elements/test_experimental/__init__.py:1]: Sounds good

Comment (frederik-sandfort1) [reply tests/test_elements/test_experimental/__init__.py:1]: But just note that in the tests you have for all others only """Init.""" in there ;-)

File: .gitignore [hunk 3]
```diff
@@ -3,4 +3,4 @@ __pycache__
 molpipeline.egg-info/
 lib/
 build/
-
+.vscode/
```
Comment (c-w-feldmann) [inline .gitignore:6]: also remove

File: molpipeline/experimental/custom_filter.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+from molpipeline.abstract_pipeline_elements.core import InvalidInstance
+from molpipeline.abstract_pipeline_elements.core import (
+    MolToMolPipelineElement as _MolToMolPipelineElement,
+)
+from molpipeline.utils.molpipeline_types import OptionalMol, RDKitMol
+
+
+class CustomFilter(_MolToMolPipelineElement):
+    """CustomFilter with bool functions on mol."""
```
Comment (JochenSiegWork) [inline molpipeline/experimental/custom_filter.py:20]: A little more explanation would be helpful for the users. Maybe: "Filters molecules based on a custom boolean function. Elements not passing the filter will be set to `InvalidInstance`s."

File: molpipeline/experimental/custom_filter.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        n_jobs : int, optional
+            number of jobs to use, by default 1
+        uuid : str, optional
+            uuid of the element, by default None
+        """
+        super().__init__(name=name, n_jobs=n_jobs, uuid=uuid)
+        self.func = func
+
+    def pretransform_single(self, value: RDKitMol) -> OptionalMol:
+        """Pretransform single value.
```
Comment (JochenSiegWork) [inline molpipeline/experimental/custom_filter.py:46]: A bit more documentation would be nice

File: molpipeline/mol2any/mol2bool.py [hunk 6]
```diff
@@ -50,3 +50,23 @@ def pretransform_single(self, value: Any) -> bool:
...
+
+        Parameters
+        ----------
+        value: Any
+            Current representation of the molecule. (Eg. SMILES, RDKit Mol, ...)
+
+        Returns
+        -------
+        Any
+            New representation of the molecule. (Eg. SMILES, RDKit Mol, Descriptor-Vector, ...)
```
Comment (JochenSiegWork) [inline molpipeline/mol2any/mol2bool.py:67]: This doc-string also is too general. What will be returned is either the input molecule or an InvalidInstance. This should be mentioned here and not SMILES or descriptor vectors.

File: tests/test_elements/test_experimental/test_custom_filter.py
Comment (JochenSiegWork) [inline tests/test_elements/test_experimental/test_custom_filter.py:1]: The path in the test directory should be the same as in the package. So this should be `tests/test_experimental/test_custom_filter.py`

**Review by JochenSiegWork — CHANGES_REQUESTED** [review]
Very nice. I made only some small comments mostly to the documentation.

File: molpipeline/experimental/custom_filter.py (hunk 4 above)
Comment (frederik-sandfort1) [reply molpipeline/experimental/custom_filter.py:20]: done

File: molpipeline/mol2any/mol2bool.py (hunk 6 above)
Comment (frederik-sandfort1) [reply molpipeline/mol2any/mol2bool.py:67]: The second part of your comment is actually not true. What will be returned is the transformed (and finalized) value or the transformed value if it is an InvalidInstance (which will likely never be reached. I will actually also remove this line and state that the transformed output value will be returned.

Comment (frederik-sandfort1) [reply molpipeline/mol2any/mol2bool.py:67]: Changed a few descriptions here. Please have a look

File: tests/test_elements/test_experimental/test_custom_filter.py
Comment (frederik-sandfort1) [reply tests/test_elements/test_experimental/test_custom_filter.py:1]: done

File: .gitignore [hunk 7]
```diff
@@ -3,4 +3,3 @@ __pycache__
 molpipeline.egg-info/
 lib/
 build/
-
```
Comment (JochenSiegWork) [inline .gitignore:6]: accidental line deletion

File: molpipeline/experimental/custom_filter.py (hunk 5 above)
Comment (JochenSiegWork) [reply molpipeline/experimental/custom_filter.py:46]: For example:

"""Pretransform single value.

Applies the custom boolean function to the molecule.

...
"""

**Review by JochenSiegWork — APPROVED** [review]
After these two minor changes it can be merged

**Review by c-w-feldmann — APPROVED** [review]

**Review by JochenSiegWork — APPROVED** [review]
ready for merge 👍

**Pull request merged**